sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -508,9 +508,11 @@ class Dataset[T] private[sql](
    * @group basic
    * @since 3.4.0
    */
-  def to(schema: StructType): DataFrame = withPlan {
-    val replaced = CharVarcharUtils.failIfHasCharVarchar(schema).asInstanceOf[StructType]
-    Project.matchSchema(logicalPlan, replaced, sparkSession.sessionState.conf)
+  def to(schema: StructType): DataFrame = withOrigin() {
```

Comment:
I think it's good enough to attach expression call site for ansi mode, we can attach plan call site later.

Reply:
Could you elaborate a little bit more, why ansi mode is important here?

Reply:
When I revert changes in method like this `DataFrame -> DataFrame`, the fragment becomes blurry, like:
Before: `select`
After: `anonfun$select$4`

Reply:
@cloud-fan For example, **with** `withOrigin` in `select`, we stop at index 3:
<img width="853" alt="Screenshot 2023-11-01 at 21 25 06" src="https://github.com/apache/spark/assets/1580697/7080a1bb-7128-4c23-baaf-64aab6f7cd0a">

Without `withOrigin` in `select`, the picture is different. We are in `withOrigin` in `Column`'s constructor, and stoped at the index 5:
<img width="862" alt="Screenshot 2023-11-01 at 21 29 18" src="https://github.com/apache/spark/assets/1580697/c6747280-fc30-4500-a26c-e0a063544966">